dandiapi/api/services/dandiset/__init__.py

```diff
@@ -76,6 +81,12 @@ def delete_dandiset(*, user, dandiset: Dandiset) -> None:
         # chance to grab the Dandiset information before it is destroyed.
         audit.delete_dandiset(dandiset=dandiset, user=user)
 
+        # Dandisets with published versions cannot be deleted
+        # so only the Dandiset DOI needs to be deleted.
+        draft_version = dandiset.versions.filter(version='draft').first()
+        if draft_version and draft_version.doi is not None:
```

Comment:
I have a couple of suggestions. First, we know a draft version will always exist on a dandiset, as it is created in the same transaction as the dandiset. So we don't need to check that it exists here.

Second, I actually think we should prevent deleting the dandiset until the DOI is generated. Imagine someone creates and then immediately deletes a dandiset, at the same time the celery task to create the DOI is running. This code block would be skipped, and the dandiset would still be deleted, but the DOI could have been created during that time. In that case, we would have "orphaned DOIs" to worry about.

I think it's acceptable to block deletion on draft DOI creation, and raise an appropriate error message from the request.